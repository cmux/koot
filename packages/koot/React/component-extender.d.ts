/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line @typescript-eslint/triple-slash-reference
/// <reference path="../global.d.ts" />

import {
    // ComponentType,
    ReactNode,
    FC,
    ComponentClass,
    Component,
    ComponentState,
    RefObject,
} from 'react';
import { Store, Dispatch } from 'redux';
import {
    MapStateToPropsParam,
    MapDispatchToPropsNonObject,
    MapDispatchToPropsParam,
    MergeProps,
    Options as ReactReduxOptions,
} from 'react-redux';

// ============================================================================

declare module 'kootExtendHOC';

// ============================================================================

interface ExtendComponent<P, S> {
    (wrappedComponent: FC<ExtendedProps & P>): ComponentClass<
        ExtendedProps & P
    >;
    (wrappedComponent: ComponentClass<any, any>): HOC<ExtendedProps & P, S>;
    // (wrappedComponent: FC<ExtendedProps & P>): ComponentClass<
    //     ExtendedProps & P
    // >;

    // (
    //     wrappedComponent:
    //         | FC<ExtendedProps & P>
    //         | ComponentClass
    //         | ComponentClass<ExtendedProps & P>
    // ): ComponentClass<ExtendedProps & P>;

    // (wrappedComponent: ComponentType<any>): HOC<ExtendedProps & P, S>;
    // (wrappedComponent: ComponentClass<any, any>): ComponentClass<
    //     JSX.LibraryManagedAttributes<wrappedComponent, P> & ExtendedProps & P,
    //     S
    // >;
}
class HOC extends Component {}

export function extend<P = {}, S = ComponentState>(options: {
    // connect?: Connect;
    connect?:
        | true
        | MapStateToPropsParam<any, P, S>
        | [null | undefined | MapStateToPropsParam<any, P, S>]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              )
          ]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              ),
              null | undefined | MergeProps<any, any, P, any>
          ]
        | [
              null | undefined | MapStateToPropsParam<any, P, S>,
              (
                  | null
                  | undefined
                  | MapDispatchToPropsNonObject<any, P>
                  | MapDispatchToPropsParam<any, P>
              ),
              null | undefined | MergeProps<any, any, P, any>,
              null | undefined | ReactReduxOptions<S, any, P>
          ];
    /** 提供页面的 title 和 meta 标签信息 */
    pageinfo?: Pageinfo | ExtendPageinfoFunction;
    data?: DataFetchFunction | ExtendData;
    styles?: KootComponentStyleObject;
    /**
     * 控制 SSR 行为
     * @default true
     */
    ssr?: ReactNode | boolean;
    /** 组件名 (仅供调试用) */
    name?: string;
}): ExtendComponent<P, S>;

export type Pageinfo = {
    /** 要设置的页面标题 */
    title?: string;
    /** 要设置的 `<meta>` 标签列表，需要元素为 `{[name]: value}` 对象的数组 */
    metas?: Array<MetaObject>;
};
export type MetaObject = {
    [metaKey: string]: string | undefined;
};

type ExtendPageinfoFunction = (state: any, props: any) => Pageinfo;

type DataFetchFunction = (
    store: Store,
    renderProps: any,
    dispatch: Dispatch
) => Promise<any>;

type DataCheckFunction = (state: any, renderProps: any) => boolean;

type ExtendData = {
    fetch?: DataFetchFunction;
    check?: DataCheckFunction;
};

//

/** extend 高阶组件向目标组件注入的 props */
export interface ExtendedProps {
    /** 包含父组件传入的 `className` 和 `extend()` 高阶组件注入的 `className`（如果在 `extend()` 高阶组件中传入了 `styles` 属性）*/
    className?: string;
    /**
     * `extend()` 高阶组件注入的 `className`
     * - 不包含父组件传入的 `className`
     * - 仅在 `extend()` 高阶组件中传入了 `styles` 属性时才会存在
     */
    readonly 'data-class-name'?: string;
    /**
     * 可用来手动触发页面信息更新，语法与 `connect()` 高阶组件的 `pageinfo` 的函数用法相同
     * - 如果没有传入参数，会使用 `extend()` 高阶组件中的 `pageinfo` 的值
     * - 仅在 `extend()` 高阶组件中使用 `pageinfo` 属性时才会存在
     */
    readonly updatePageinfo?: ExtendedPropsUpdatePageinfo;
    readonly forwardedRef?: RefObject<any>;
}
type ExtendedPropsUpdatePageinfo = (
    newPageinfo?: Pageinfo | ExtendPageinfoFunction
) => void;

interface RenderProps extends ExtendedProps {
    readonly location?: Record<string, any>;
    readonly params?: Record<string, any>;
    readonly route?: Record<string, any>;
    readonly routeParams?: Record<string, any>;
    readonly router?: Record<string, any>;
    readonly routes?: Array<Record<string, any>>;
}
