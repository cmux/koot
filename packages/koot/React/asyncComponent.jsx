/* eslint-disable react/no-unsafe */
// https://gist.github.com/acdlite/a68433004f9d6b4cbc83b5cc3990c194

const { Component } = require('react');

module.exports = (getComponent) =>
    class AsyncComponent extends Component {
        static Component = null;
        state = { Component: AsyncComponent.Component };

        componentWillMount() {
            if (!this.state.Component) {
                getComponent().then((Component) => {
                    AsyncComponent.Component = Component;
                    this.setState({ Component });
                });
            }
        }
        render() {
            const { Component } = this.state;
            if (Component) {
                return <Component {...this.props} />;
            }
            return null;
        }
    };
