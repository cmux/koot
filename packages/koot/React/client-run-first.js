import 'regenerator-runtime/runtime';
import { syncHistoryWithStore } from 'react-router-redux';
import * as portionConfig from '__KOOT_PROJECT_CONFIG_PORTION_CLIENT_PATHNAME__';
import { STORE, HISTORY, DEV_NATIVE_CONSOLE } from '../defaults/defines-window';
import validateReduxConfig from './validate/redux-config';
import History from './history';

window[STORE] = ((reduxConfig = {}) => {
    if (typeof reduxConfig.factoryStore === 'function')
        return reduxConfig.factoryStore();
    if (typeof reduxConfig.store === 'object') return reduxConfig.store;
    return {};
})(validateReduxConfig(portionConfig.redux));

window[HISTORY] = syncHistoryWithStore(History, window[STORE]);

// console filter
// https://stackoverflow.com/questions/6659312/is-there-a-way-to-filter-output-in-google-chromes-console
if (__DEV__) {
    (() => {
        // eslint-disable-next-line no-console
        console.info('ðŸ’• [DEV] Thanks for using Koot.js');

        window.console = window.console || { log: 0 };

        const nativeConsole = {
            log: window.console.log,
            info: window.console.info,
            error: window.console.error,
        };
        window[DEV_NATIVE_CONSOLE] = nativeConsole;

        const lastLog = {
            WDS: {
                time: undefined,
                log: undefined,
            },
            HMR: {
                time: undefined,
                log: undefined,
            },
        };
        const throttleHotLog = (level) => {
            window.console[level] = function () {
                try {
                    const args = Array.from(arguments);

                    // throttle `WDS` and `HMR` logs
                    const needThrottle = (type) => {
                        const regex = new RegExp(`^\\[${type}\\]`);
                        if (!args.some((arg) => regex.test(arg))) return false;

                        const { [type]: last } = lastLog;
                        const now = Date.now();
                        const nowLogStr = args.join('|||');
                        const { time: lastTime, log: lastLogStr } = last;
                        lastLog[type] = {
                            time: now,
                            log: nowLogStr,
                        };
                        // console.warn(type, last, now - lastTime)

                        if (
                            lastTime &&
                            now - lastTime < 50 &&
                            lastLogStr &&
                            lastLogStr === nowLogStr
                        )
                            return true;
                        return false;
                    };
                    if (needThrottle('WDS')) return;
                    if (needThrottle('HMR')) return;

                    nativeConsole[level].apply(window.console, arguments);
                } catch (e) {}
            };
        };
        // throttleHotLog('log')
        throttleHotLog('info');

        const warningShowed = {
            reactRouterV3: false,
            reactReduxV5: false,
        };
        window.console.error = function () {
            try {
                const args = Array.from(arguments);

                // filter out SSR not matching error
                // will fix this by using koot's own async component
                // SOONâ„¢
                if (
                    args.some((arg) =>
                        /^Warning: Did not expect server HTML to contain a <%s> in <%s>\.$/.test(
                            arg
                        )
                    )
                )
                    return;

                // filter out react-router v3` warnings/errors
                if (
                    (/^Warning: Legacy context API has been detected within a strict-mode tree/.test(
                        args[0]
                    ) &&
                        args.some((arg) => arg === 'Link')) ||
                    (/^Warning: Using UNSAFE_[a-zA-Z]+ in strict mode/.test(
                        args[0]
                    ) &&
                        /(Link)(\r|\n|,|\(|$)/.test(args[1]))
                ) {
                    if (!warningShowed.reactRouterV3) {
                        // eslint-disable-next-line no-console
                        console.info(
                            '\n[koot] Koot.js is now using `react-router` v3 which will be upgraded to newer version in future.\n\n'
                        );
                        warningShowed.reactRouterV3 = true;
                    }
                    return;
                }

                nativeConsole.error.apply(window.console, arguments);
            } catch (e) {}
        };
    })();
}
