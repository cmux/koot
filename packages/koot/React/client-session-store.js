import get from 'lodash/get';
import set from 'lodash/set';

import { getStore } from '../';
import { sessionStoreKey } from '../defaults/defines';
import filterState from '../libs/filter-state';

const configSessionStore = JSON.parse(process.env.KOOT_SESSION_STORE);

/** @type {Boolean} 当前是否可以/允许使用 sessionStore */
const able = (() => {
    if (__SERVER__) return false;
    if (!window.sessionStorage) return false;
    if (!configSessionStore) return false;

    if (configSessionStore === true) return true;
    if (configSessionStore === 'all') return true;

    return Boolean(
        typeof configSessionStore === 'object' &&
            !Array.isArray(configSessionStore)
    );
})();

/**
 * 保存当前 state 到 sessionStorage
 * @void
 */
export const save = () => {
    if (!able) return;

    /** @type {Object} 排除掉黑名单内的项目后的 state 对象 */
    const state = filterState(getStore().getState());

    let saveState = {};

    if (configSessionStore === true || configSessionStore === 'all') {
        saveState = state;
    } else if (
        typeof configSessionStore === 'object' &&
        !Array.isArray(configSessionStore)
    ) {
        // 根据配置对象存储
        const parse = (obj, accumulatedKey = '') => {
            Object.keys(obj).forEach((key) => {
                const currentAccumulatedKey =
                    accumulatedKey + `[${JSON.stringify(key)}]`;
                if (typeof obj[key] === 'object') {
                    parse(obj[key], currentAccumulatedKey);
                } else if (obj[key] === true) {
                    const value = get(state, currentAccumulatedKey);
                    if (typeof value !== 'undefined')
                        set(saveState, currentAccumulatedKey, value);
                }
            });
        };
        parse(configSessionStore);
        // console.log(configSessionStore, state, saveState);
    }

    sessionStorage.setItem(sessionStoreKey, JSON.stringify(saveState));

    return;
};

/**
 * 向 window.onunload 添加事件：保存 state
 * @void
 */
export const addEventHandlerOnPageUnload = () => {
    if (!able) return;
    window.addEventListener('unload', save);
    return;
};

/**
 * 从 sessionStorage 中读取 state
 * @returns {Object} 存储的 state
 */
export const load = () => {
    if (!able) return {};
    return JSON.parse(sessionStorage.getItem(sessionStoreKey) || '{}');
};
