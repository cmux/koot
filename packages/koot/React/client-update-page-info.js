/**
 * @typedef Pageinfo
 * @type {Object}
 * @property {string} [title] 标题
 * @property {Array<Object.<string, string>>} [metas] meta 标签信息，需要形式为 `{[name]: value}` 对象的数组
 */

// import isEqual from 'lodash/isEqual';

/**
 * 当前已注入的 meta 标签
 * @type {Array<HTMLElement>}
 */
let injectedMetaTags;

// meta 标签区域结尾的 HTML 注释代码
let nodeCommentEnd;

let inited = false;

/** @type {Pageinfo} */
const infoToChange = {
    title: undefined,
    metas: [],
};
let changeTimeout = undefined;

/**
 * _仅限客户端_
 *
 * 更新页面信息
 * @module
 * @param {string} title 新的标题
 * @param {Array<Object.<string, string>>} metas meta 标签信息，需要形式为 `{[name]: value}` 对象的数组
 */
export default (title, metas = []) => {
    if (__SERVER__) return;
    if (!__SPA__ && !inited) {
        setTimeout(() => {
            inited = true;
        });
        return;
    }

    // 判断 & 追加即将修改的内容
    /*
    // 如果新 meta 不存在于已有的 meta 列表，添加
    metas.forEach(metaNew => {
        if (!infoToChange.metas.length) infoToChange.metas.push(metaNew);
        if (
            !infoToChange.metas.every(metaExist => {
                console.log(metaNew, metaExist, isEqual(metaNew, metaExist));
                return isEqual(metaNew, metaExist);
            })
        ) {
            console.log('PUSH');
            infoToChange.metas.push(metaNew);
        }
    });
    */
    // 如果当前没有信息，设为当前信息
    if (
        !infoToChange.title &&
        (!Array.isArray(infoToChange.metas) || !infoToChange.metas.length)
    ) {
        infoToChange.title = title;
        infoToChange.metas = metas;
    }

    if (changeTimeout) return;

    changeTimeout = setTimeout(() => {
        doUpdate();
        infoToChange.title = undefined;
        infoToChange.metas = [];
        changeTimeout = undefined;
    });
};

//

const doUpdate = () => {
    const { title, metas } = infoToChange;

    // 替换页面标题
    if (typeof title !== 'undefined') document.title = title;

    // 替换 metas
    const head = document.getElementsByTagName('head')[0];
    getInjectedMetaTags().forEach((el) => head.removeChild(el));

    injectedMetaTags.forEach((el) => {
        if (el && el.parentNode) el.parentNode.removeChild(el);
    });
    injectedMetaTags = metas
        .filter((meta) => typeof meta === 'object')
        .map((meta) => {
            const el = document.createElement('meta');
            for (var key in meta) {
                el.setAttribute(key, meta[key]);
            }
            // el.setAttribute(__KOOT_INJECT_ATTRIBUTE_NAME__, '')
            if (nodeCommentEnd) {
                head.insertBefore(el, nodeCommentEnd);
            } else {
                head.appendChild(el);
            }
            return el;
        });
};

//

/**
 * 获取当前已注入的 meta 标签
 * @returns {Array<HTMLElement>}
 */
export const getInjectedMetaTags = () => {
    if (!Array.isArray(injectedMetaTags)) {
        const head = document.getElementsByTagName('head')[0];

        injectedMetaTags = [];
        // 移除所有在 KOOT_METAS 里的 meta 标签
        // 采用 DOM 操作的初衷：如果使用 innerHTML 的字符串替换方法，浏览器可能会全局重新渲染一次，造成“闪屏”
        const childNodes = head.childNodes;
        let meetStart = false;
        let meetEnd = false;
        let i = 0;
        while (!meetEnd && childNodes[i] instanceof Node) {
            const node = childNodes[i];
            if (node.nodeType === Node.COMMENT_NODE) {
                if (node.nodeValue === __KOOT_INJECT_METAS_START__)
                    meetStart = true;
                if (node.nodeValue === __KOOT_INJECT_METAS_END__) {
                    meetEnd = true;
                    nodeCommentEnd = node;
                }
            } else if (
                meetStart &&
                node.nodeType === Node.ELEMENT_NODE &&
                node.tagName === 'META'
            ) {
                injectedMetaTags.push(node);
            }
            i++;
        }
    }

    return injectedMetaTags;
};

/** _仅针对客户端_ 标记已初始化 */
export const markInited = () => {
    inited = true;
};
