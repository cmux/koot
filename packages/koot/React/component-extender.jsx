/* __KOOT_DEV_NO_REACT_HOT_INJECT__ */
import { Component, forwardRef } from 'react';
import { connect } from 'react-redux';
// import PropTypes from 'prop-types'
import hoistStatics from 'hoist-non-react-statics';
import md5 from 'md5';

import {
    needConnectComponents,
    ssrContext as SSRContext,
} from '../defaults/defines-server';

//

import {
    fromServerProps as getRenderPropsFromServerProps,
    fromComponentProps as getRenderPropsFromComponentProps,
} from './get-render-props';
import {
    append as appendStyle,
    remove as removeStyle,
    // StyleMapContext,
} from './styles';
import clientUpdatePageInfo from './client-update-page-info';
import { RESET_CERTAIN_STATE } from './redux';
import isRenderSafe from './is-render-safe';
import isFunctionalComponent from './is-functional-component';

import RootContext from './root-context';

//

// 是否已挂载了组件
let everMounted = false;
// const defaultPageInfo = {
//     title: '',
//     metas: []
// }

const propsNameForwardRef = 'koot' + md5('forwardedRef');

/**
 * @type {Number}
 * _开发环境_
 * _服务器_
 * 使用该高阶组件的次数
 */
let devSSRConnectIndex = 0;

/**
 * 获取数据
 * @callback callbackFetchData
 * @param {Object} state 当前 state
 * @param {Object} renderProps 封装的同构 props
 * @param {Function} dispatch Redux dispatch
 * @returns {Promise}
 */

/**
 * 判断数据是否准备好
 * @callback callbackCheckLoaded
 * @param {Object} state 当前 state
 * @param {Object} renderProps 封装的同构 props
 * @returns {boolean}
 */

/**
 * 获取页面信息
 * @callback callbackGetPageInfo
 * @param {Object} state 当前 state
 * @param {Object} renderProps 封装的同构 props
 * @returns {Object}
 */

/**
 * 获取同构数据的执行方法
 * @param {Object} store
 * @param {Object} props renderProps
 * @returns {Promise}
 */
const doFetchData = (store, renderProps, funcFetch) => {
    // return new Promise((resolve) => resolve());
    if (!isRenderSafe())
        return __CLIENT__ ? new Promise((resolve) => resolve()) : undefined;

    const result = funcFetch(store.getState(), renderProps, store.dispatch);
    // if (result === true) {
    //     isDataPreloaded = true
    //     return new Promise(resolve => resolve())
    // }
    if (Array.isArray(result)) return Promise.all(result);
    if (result instanceof Promise) return result;
    return new Promise((resolve) => resolve(result));
};

/**
 * 更新页面信息
 * @param {Object} store
 * @param {Object} props renderProps
 * @returns {Object} infos
 * @returns {String} infos.title
 * @returns {Array} infos.metas
 */
const doPageinfo = (store, props, pageinfo) => {
    if (!isRenderSafe()) return {};

    const defaultPageInfo = {
        title: undefined,
        metas: [],
    };

    if (typeof pageinfo !== 'function' && typeof pageinfo !== 'object')
        return defaultPageInfo;

    const state = store.getState();
    const infos = (() => {
        if (typeof pageinfo === 'object') return pageinfo;
        const infos = pageinfo(state, props);
        if (typeof infos !== 'object') return defaultPageInfo;
        return infos;
    })();

    const { title = defaultPageInfo.title, metas = defaultPageInfo.metas } =
        infos;

    if (state.localeId) {
        if (
            !metas.some((meta) => {
                if (meta.name === 'koot-locale-id') {
                    meta.content = state.localeId;
                    return true;
                }
                return false;
            })
        ) {
            metas.push({
                name: 'koot-locale-id',
                content: state.localeId,
            });
        }
    }

    return {
        title,
        metas,
    };
};

// console.log((typeof store === 'undefined' ? `\x1b[31m×\x1b[0m` : `\x1b[32m√\x1b[0m`) + ' store in [HOC] extend')
/**
 * 高阶组件/组件装饰器：组件扩展
 * @param {Object} options 选项
 * @param {boolean|Function} [options.connect] react-redux 的 connect() 的参数。如果为 true，表示使用 connect()，但不连接任何数据
 * @param {Object|callbackGetPageInfo} [options.pageinfo]
 * @param {Object} [options.data] 同构数据相关
 * @param {callbackFetchData} [options.data.fetch]
 * @param {callbackCheckLoaded} [options.data.check]
 * @param {Object} [options.styles] 组件 CSS 结果
 * @returns {Function} 封装好的 React 组件
 */
export default (options = {}) =>
    (WrappedComponent) => {
        // console.log((typeof store === 'undefined' ? `\x1b[31m×\x1b[0m` : `\x1b[32m√\x1b[0m`) + ' store in [HOC] extend run')

        const {
            connect: _connect = false,
            pageinfo,
            data: {
                fetch: _dataFetch,
                check: dataCheck,
                resetWhenUnmount: dataResetWhenUnmount,
            } = {},
            styles: _styles,
            ssr = true,
            // ttt
            // hot: _hot = true,
            // name
        } = options;

        // console.log('extend hoc run', { name, LocaleId })

        // 样式相关

        /** @type {Object} 经过 koot-css-loader 处理后的 css 文件的结果对象 */
        const styles = (!Array.isArray(_styles) ? [_styles] : _styles).filter(
            (obj) => typeof obj === 'object' && typeof obj.wrapper === 'string'
        );

        /** @type {boolean} 是否有上述结果对象 */
        const hasStyles = Array.isArray(styles) && styles.length > 0;
        // console.log({ ttt, hasStyles, styles })

        /** @type {boolean} 是否有 pageinfo 对象 */
        const hasPageinfo =
            typeof pageinfo === 'function' || typeof pageinfo === 'object';

        // 同构数据相关

        /** @type {boolean} 同构数据是否已经获取成功 */
        // let isDataPreloaded = false

        /** @type {Function} 获取同构数据 */
        const dataFetch =
            typeof options.data === 'function' || Array.isArray(options.data)
                ? options.data
                : typeof _dataFetch === 'function' || Array.isArray(_dataFetch)
                ? _dataFetch
                : undefined;

        // 装饰组件
        class KootReactComponent extends Component {
            static onServerRenderHtmlExtend = ({ store, renderProps = {} }) => {
                const { title, metas } = doPageinfo(
                    store,
                    getRenderPropsFromServerProps(renderProps),
                    pageinfo
                );
                return { title, metas };
            };

            //

            clientUpdatePageInfo(to) {
                if (!__CLIENT__) return;
                if (!hasPageinfo) return;

                const { title, metas } =
                    typeof to === 'function'
                        ? doPageinfo(
                              this.context.store,
                              this.getRenderProps(),
                              to
                          )
                        : to ||
                          doPageinfo(
                              this.context.store,
                              this.getRenderProps(),
                              pageinfo
                          );

                clientUpdatePageInfo(
                    title || process.env.KOOT_PROJECT_NAME,
                    metas
                );
            }
            getRenderProps() {
                return getRenderPropsFromComponentProps(this.props);
            }

            //

            state = {
                loaded:
                    typeof dataCheck === 'function'
                        ? dataCheck(
                              this.context.store.getState(),
                              this.getRenderProps()
                          )
                        : undefined,
            };
            mounted = false;
            kootClassNames = [];

            //

            constructor(props, context) {
                super(props, context);

                /**
                 * _服务器端_
                 * 将组件注册到同构渲染对象中
                 */
                if (__SERVER__) {
                    const { [SSRContext]: SSR } = context.ctx;
                    if (SSR[needConnectComponents]) {
                        if (__DEV__) {
                            // console.log(options.name || '__');
                            KootComponent.id = devSSRConnectIndex++;
                            // KootComponent.pageinfo = pageinfo;
                        }
                        if (Array.isArray(SSR.connectedComponents))
                            SSR.connectedComponents.unshift(KootComponent);
                    }
                }

                if (!isRenderSafe()) return;

                if (hasStyles) {
                    this.kootClassNames = styles.map((obj) => obj.wrapper);
                    appendStyle(this.getStyleMap(/*context*/), styles);
                    // console.log('----------')
                    // console.log('styles', styles)
                    // console.log('theStyles', theStyles)
                    // console.log('this.classNameWrapper', this.classNameWrapper)
                    // console.log('----------')
                }
            }

            /**
             * 获取 styleMap
             * - 服务器端: 全局常量中的对照表
             * - 客户端: 静态对象 styles
             */
            getStyleMap(/*context*/) {
                if (__SERVER__) return this.context.styles;
                return KootReactComponent.styles;
            }

            //

            componentDidUpdate(prevProps) {
                // if (
                //     typeof prevProps.location === 'object' &&
                //     typeof this.props.location === 'object' &&
                //     prevProps.location.pathname !== this.props.location.pathname
                // )
                this.clientUpdatePageInfo();
            }

            componentDidMount() {
                this.mounted = true;

                if (!this.state.loaded && typeof dataFetch !== 'undefined') {
                    doFetchData(
                        this.context.store,
                        this.getRenderProps(),
                        dataFetch
                    ).then(() => {
                        if (!this.mounted) return;
                        this.setState({
                            loaded: true,
                        });
                    });
                }

                this.clientUpdatePageInfo();
                if (hasPageinfo && this.mounted) {
                    setTimeout(() => {
                        if (this && this.mounted) {
                            this.clientUpdatePageInfo();
                        }
                    }, 500);
                }

                if (everMounted) {
                } else {
                    everMounted = true;
                }
            }

            componentWillUnmount() {
                this.mounted = false;
                if (hasStyles) {
                    removeStyle(this.getStyleMap(/*this.context*/), styles);
                }
                if (typeof dataResetWhenUnmount === 'object') {
                    setTimeout(() => {
                        this.props.dispatch({
                            type: RESET_CERTAIN_STATE,
                            data: dataResetWhenUnmount,
                        });
                    });
                }
            }

            //

            render() {
                // console.log('styles', styles)
                // console.log('this', this)
                // console.log('this.kootClassNames', this.kootClassNames)
                // console.log('this.props.className', this.props.className)

                if (__SERVER__) {
                    if (ssr === false) return null;
                    if (ssr !== true) return ssr;
                }

                if (__CLIENT__ && this.kootClassNames instanceof HTMLElement) {
                    // console.log(this.kootClassNames)
                    this.kootClassNames = [
                        this.kootClassNames.getAttribute('id'),
                    ];
                }

                const props = Object.assign({}, this.props, {
                    className: this.kootClassNames
                        .concat(this.props.className)
                        .join(' ')
                        .trim(),
                    'data-class-name': this.kootClassNames.join(' ').trim(),
                });
                if (hasPageinfo)
                    props.updatePageinfo = this.clientUpdatePageInfo.bind(this);

                // if (__SERVER__) console.log('extender this.state.loaded', this.state.loaded)
                if (
                    typeof dataFetch !== 'undefined' &&
                    typeof dataCheck === 'function'
                )
                    props.loaded = this.state.loaded;

                // if (typeof props.forwardedRef !== 'undefined') {
                //     console.log(props.forwardedRef);
                // }
                if (typeof props[propsNameForwardRef] !== 'undefined') {
                    props.forwardedRef = props[propsNameForwardRef];
                }
                if (
                    !isFunctionalComponent(WrappedComponent) &&
                    !props.ref &&
                    (process.env.KOOT_REACT_LEGACY_REF ||
                        !props.forwardedRef ||
                        !Object.keys(props.forwardedRef).includes('current'))
                ) {
                    props.ref =
                        props[propsNameForwardRef] ??
                        props.forwardedRef ??
                        props.ref;
                }
                delete props[propsNameForwardRef];

                return <WrappedComponent {...props} />;
            }
        }

        KootReactComponent.contextType = RootContext;
        KootReactComponent.styles = {};

        if (typeof dataFetch !== 'undefined') {
            KootReactComponent.onServerRenderStoreExtend = ({
                store,
                renderProps,
            }) =>
                doFetchData(
                    store,
                    getRenderPropsFromServerProps(renderProps),
                    dataFetch
                );
        }

        // if (typeof styles === 'object' &&
        //     typeof styles.wrapper === 'string'
        // ) {
        //     KootComponent = ImportStyle(styles)(KootComponent)
        // }
        let KootComponent = hoistStatics(KootReactComponent, WrappedComponent);

        // if (typeof styles === 'object' &&
        //     typeof styles.wrapper === 'string'
        // ) {
        //     KootComponent = ImportStyle(styles)(KootComponent)
        // }

        if (_connect === true) {
            KootComponent = connect(() => ({}), undefined, undefined, {
                forwardRef: true,
            })(KootComponent);
        } else if (typeof _connect === 'function') {
            KootComponent = connect(_connect, undefined, undefined, {
                forwardRef: true,
            })(KootComponent);
        } else if (Array.isArray(_connect)) {
            if (typeof _connect[3] !== 'object') _connect[3] = {};
            _connect[3].forwardRef = true;
            KootComponent = connect(..._connect)(KootComponent);
        }

        // if (isFunctionalComponent(KootComponent)) return KootComponent;

        // return KootComponent;
        return forwardRef((props, ref) => {
            const thisProps = {
                ...props,
                [propsNameForwardRef]: ref ?? props.forwardedRef,
            };
            return <KootComponent {...thisProps} />;
        });
    };
