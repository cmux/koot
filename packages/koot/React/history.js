// import history from "__KOOT_CLIENT_REQUIRE_HISTORY__"
import createHistory from '__KOOT_CLIENT_REQUIRE_CREATE_HISTORY__';
import { parsePath } from 'history/lib/PathUtils';
import isI18nEnabled from '../i18n/is-enabled';
import { REDUXSTATE } from '../defaults/defines-window';

/**
 * History Enhancer: use basename
 *
 * Original useBasename enhancer from history also override all read methods
 * `getCurrentLocation` `listenBefore` `listen`
 * But as Diablohu tested, when read methods overrided, if the route matched used async method to get component, would fail
 * that rendering blank page and no route match event fired
 * So we only overrid write methods here. And modify the first level path in routes object to `:localeId`
 *
 * @param {Function} createHistory
 * @returns {Object} History
 */
const kootUseBasename =
    (createHistory) =>
    (options = {}) => {
        const history = createHistory(options);
        const { basename } = options;

        const addBasename = (location) => {
            if (!location) return location;

            if (basename && location.basename == null) {
                if (
                    location.pathname
                        .toLowerCase()
                        .indexOf(basename.toLowerCase()) === 0
                ) {
                    location.pathname = location.pathname.substring(
                        basename.length
                    );
                    location.basename = basename;

                    if (location.pathname === '') location.pathname = '/';
                } else {
                    location.basename = '';
                }
            }

            return location;
        };

        const prependBasename = (location) => {
            if (!basename) return location;

            const object =
                typeof location === 'string' ? parsePath(location) : location;
            const pname = object.pathname;
            const normalizedBasename =
                basename.slice(-1) === '/' ? basename : `${basename}/`;
            const normalizedPathname =
                pname.charAt(0) === '/' ? pname.slice(1) : pname;
            const pathname = normalizedBasename + normalizedPathname;

            return {
                ...object,
                pathname,
            };
        };

        // Override all write methods with basename-aware versions.
        const push = (location) => history.push(prependBasename(location));

        const replace = (location) =>
            history.replace(prependBasename(location));

        const createPath = (location) =>
            history.createPath(prependBasename(location));

        const createHref = (location) =>
            history.createHref(prependBasename(location));

        const createLocation = (location, ...args) =>
            addBasename(
                history.createLocation(prependBasename(location), ...args)
            );

        return {
            ...history,
            push,
            replace,
            createPath,
            createHref,
            createLocation,
        };
    };

let historyClient;

const history = (() => {
    if (__CLIENT__) {
        if (!historyClient) {
            const initialState = window[REDUXSTATE] || {};
            const historyConfig = { basename: '/' };
            if (
                isI18nEnabled() &&
                process.env.KOOT_I18N_URL_USE === 'router' &&
                initialState.localeId
            ) {
                historyConfig.basename = `/${initialState.localeId}`;
                historyClient = kootUseBasename(createHistory)(historyConfig);
            } else if (!!process.env.KOOT_HISTORY_BASENAME) {
                const basename = process.env.KOOT_HISTORY_BASENAME.replace(
                    /^\//,
                    ''
                ).replace(/\/$/, '');
                historyConfig.basename = `/${basename}`;
                historyClient = kootUseBasename(createHistory)(historyConfig);
            } else {
                // historyClient = require("__KOOT_CLIENT_REQUIRE_HISTORY__")
                historyClient = createHistory();
            }
            // console.log({
            //     historyConfig,
            //     historyClient
            // })
        }
        return historyClient;
    }

    if (__SERVER__) {
        return undefined;
    }
})();

export default history;
