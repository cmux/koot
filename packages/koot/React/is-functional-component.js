/** 给定的组件是否为函数组件 */
function isFunctionalComponent(Component) {
    return (
        typeof Component === 'function' && // can be various things
        !(
            (
                Component.prototype && // native arrows don't have prototypes
                Component.prototype.isReactComponent
            ) // special property
        )
    );
}

export default isFunctionalComponent;
