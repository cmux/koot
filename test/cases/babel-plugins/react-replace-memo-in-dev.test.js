const babel = require('@babel/core');
const plugin = require('../../../packages/koot-webpack/loaders/babel/plugins/react-replace-memo-in-dev');

const examples = {
    test1: `
import React from 'react';
const A = React.memo(() => 'abc');
`,
    test2: `
import React2 from 'react';
const A = React2.memo(() => 'abc');
`,
    test3: `
import React, { memo } from 'react';
const A = memo(() => 'abc');
`,
    test4: `
import React, { memo as memo2 } from 'react';
const A = memo2(() => 'abc');
`,
};

describe('Babel Plugin: react-replace-memo-in-dev', () => {
    for (const [name, codeOriginal] of Object.entries(examples)) {
        it(name, () => {
            const { code } = babel.transform(codeOriginal, {
                plugins: [plugin],
            });
            expect(code).toMatchSnapshot();
        });
    }
});
