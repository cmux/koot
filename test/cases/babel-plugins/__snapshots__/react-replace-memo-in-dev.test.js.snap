// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Babel Plugin: react-replace-memo-in-dev test1 1`] = `
"import React from 'react';
const A = () => 'abc';"
`;

exports[`Babel Plugin: react-replace-memo-in-dev test2 1`] = `
"import React2 from 'react';
const A = () => 'abc';"
`;

exports[`Babel Plugin: react-replace-memo-in-dev test3 1`] = `
"import React, { memo } from 'react';
const A = () => 'abc';"
`;

exports[`Babel Plugin: react-replace-memo-in-dev test4 1`] = `
"import React, { memo as memo2 } from 'react';
const A = () => 'abc';"
`;
